crates/bevy_pbr/src/decal/projector.rs

```diff
@@ -0,0 +1,364 @@
+use bevy_ecs::{
+    component::{require, Component},
+    entity::{Entity, EntityHashMap},
+    prelude::ReflectComponent,
+    query::With,
+    schedule::IntoSystemConfigs as _,
+    system::{Query, Res, ResMut, Resource},
+};
+use bevy_image::Image;
+use bevy_math::Mat4;
+use bevy_reflect::Reflect;
+use bevy_render::{
+    extract_component::{ExtractComponent, ExtractComponentPlugin},
+    render_asset::RenderAssets,
+    render_resource::{
+        binding_types, BindGroupLayoutEntryBuilder, Buffer, BufferUsages, RawBufferVec, Sampler,
+        SamplerBindingType, Shader, ShaderType, TextureSampleType, TextureView,
+    },
+    renderer::{RenderAdapter, RenderDevice, RenderQueue},
+    sync_world::RenderEntity,
+    texture::{FallbackImage, GpuImage},
+    view::{self, ViewVisibility, Visibility, VisibilityClass},
+    Extract, ExtractSchedule, Render, RenderApp, RenderSet,
+};
+use bevy_transform::{components::GlobalTransform, prelude::Transform};
+use bevy_utils::HashMap;
+use bytemuck::{Pod, Zeroable};
+
+use crate::{
+    binding_arrays_are_usable, prepare_lights, GlobalClusterableObjectMeta, LightVisibilityClass,
+};
+
+/// The handle to the `clustered.wgsl` shader.
+pub(crate) const CLUSTERED_DECAL_SHADER_HANDLE: Handle<Shader> =
+    Handle::weak_from_u128(2881025580737984685);
+
+/// The maximum number of decals that can be present in a view.
+///
+/// This number is currently relatively low in order to work around the lack of
+/// first-class binding arrays in `wgpu`. When that feature is implemented, this
+/// limit can be increased.
+pub(crate) const MAX_VIEW_DECALS: usize = 16;
+
+/// A plugin that adds support for projected decals.
+///
+/// In environments where bindless textures aren't available, decal projectors
+/// can still be added to a scene, but they won't project any decals.
+pub struct DecalProjectorPlugin;
+
+/// An object that projects a decal onto surfaces within its bounds.
+///
+/// Conceptually, a decal projector is a 1×1×1 cube centered on its origin. It
+/// projects the given [`Self::image`] onto surfaces in the +Z direction (thus
+/// you may find [`Transform::looking_at`] useful).
+///
+/// Clustered decals are the highest-quality types of decals that Bevy supports,
+/// but they require bindless textures. This means that they presently can't be
+/// used on WebGL 2, WebGPU, macOS, or iOS. Bevy's clustered decals can be used
+/// with forward or deferred rendering and don't require a prepass.
+#[derive(Component, Debug, Clone, Reflect, ExtractComponent)]
+#[reflect(Component, Debug)]
+#[require(Transform, Visibility, VisibilityClass)]
+#[component(on_add = view::add_visibility_class::<LightVisibilityClass>)]
+pub struct DecalProjector {
+    /// The image that the decal projector projects.
+    ///
+    /// This must be a 2D image. If it has an alpha channel, it'll be alpha
+    /// blended with the underlying surface and/or other decals. All images in
+    /// the scene must use the same sampler.
```

Comment:
```suggestion
    /// blended with the underlying surface and/or other decals. All images in
    /// the scene must use the same sampler.
```

Please reword this so that it's clear that all images used for _decals_ in the scene must use the same sampler (descriptor?), but other images can use other samplers.

Reply:
Good catch, done.